File: OpenIM-SDK/src/main/java/io/openim/android/sdk/utils/JsonUtil.java
package io.openim.android.sdk.utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class JsonUtil {

    private static final Gson gson = new Gson();

    /**
     * Convert an object to a JSON string
     */
    public static String toString(Object object) {
        String gsonString = "";
        try {
            if (null != object) {
                gsonString = gson.toJson(object);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return gsonString;
    }

    /**
     * Convert an object to a JSON string without null values
     */
    public static String toStringWithoutNull(Object object) {
        String gsonString = "";
        try {
            if (null != object) {
                gsonString = new GsonBuilder().create().toJson(object);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return gsonString;
    }


    /**
     * Convert a JSON string to a generic bean
     */
    public static <T> T toObj(String gsonString, Class<T> cls) {
        T t = null;
        try {
            if (null != gsonString) {
                t = gson.fromJson(gsonString, cls);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return t;
    }


    /**
     * Convert to list
     * Generic type erasure at compile time causes errors
     * @param gsonString JSON string
     * @param cls Class type
     * @return List of type T
     */
//    public static <T> List<T> GsonToList(String gsonString, Class<T> cls) {
//        List<T> list = null;
//        if (gson != null) {
//            list = gson.fromJson(gsonString, new TypeToken<List<T>>() {
//            }.getType());
//        }
//        return list;
//    }

    /**
     * Convert to list and resolve generic type erasure issue
     */
    public static <T> List<T> toArray(String json, Class<T> cls) {
        List<T> list = null;
        try {
            if (null != json) {
                list = new ArrayList<T>();
                JsonArray array = JsonParser.parseString(json).getAsJsonArray();
                for (final JsonElement elem : array) {
                    list.add(gson.fromJson(elem, cls));
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return list;
    }


    /**
     * Convert to a list containing maps
     */
    public static <T> List<Map<String, T>> toListMaps(String gsonString) {
        List<Map<String, T>> list = null;
        if (null != gsonString) {
            list = gson.fromJson(gsonString,
                new TypeToken<List<Map<String, T>>>() {
                }.getType());
        }
        return list;
    }


    /**
     * Convert to a map
     */
    public static <T> Map<String, T> toMaps(String gsonString) {
        Map<String, T> map = null;
        if (null != gsonString) {
            map = gson.fromJson(gsonString, new TypeToken<Map<String, T>>() {
            }.getType());
        }
        return map;
    }
}

File: OpenIM-SDK/src/main/java/io/openim/android/sdk/utils/JsonUtil.kt
package io.openim.android.sdk.utils

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonArray
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken

object JsonUtil {
    private val gson: Gson = Gson()

    fun toString(obj: Any?): String {
        var gsonString = ""
        try {
            if (obj != null) {
                gsonString = gson.toJson(obj)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return gsonString
    }

    fun toStringWithoutNull(obj: Any?): String {
        var gsonString = ""
        try {
            if (obj != null) {
                gsonString = GsonBuilder().create().toJson(obj)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return gsonString
    }

    fun <T> toObj(gsonString: String?, cls: Class<T>): T? {
        var t: T? = null
        try {
            if (gsonString != null) {
                t = gson.fromJson(gsonString, cls)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return t
    }

    fun <T> toArray(json: String?, cls: Class<T>): List<T> {
        val list = ArrayList<T>()
        try {
            if (json != null) {
                val array: JsonArray = JsonParser.parseString(json).asJsonArray
                for (elem in array) {
                    list.add(gson.fromJson(elem, cls))
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return list
    }

    fun <T> toListMaps(gsonString: String?): List<Map<String, T>>? {
        return if (gsonString != null) gson.fromJson(gsonString, object : TypeToken<List<Map<String, T>>>() {}.type) else null
    }

    fun <T> toMaps(gsonString: String?): Map<String, T>? {
        return if (gsonString != null) gson.fromJson(gsonString, object : TypeToken<Map<String, T>>() {}.type) else null
    }
}
